src/storage.rs

```diff
@@ -105,23 +111,53 @@ where
     }
 }
 
+/// Implement your own token storage solution by implementing this trait. You need a way to
+/// store and retrieve tokens, each keyed by a set of scopes.
+#[async_trait]
+pub trait TokenStorage: Send + Sync {
+    /// Store a token for the given set of scopes so that it can be retrieved later by get()
+    /// ScopeSet implements Hash so that you can easily serialize and store it.
+    /// TokenInfo can be serialized with serde.
+    async fn set(&self, scopes: &[&str], token: TokenInfo) -> anyhow::Result<()>;
+
+    /// Retrieve a token stored by set for the given set of scopes
+    async fn get(&self, scopes: &[&str]) -> Option<TokenInfo>;
+}
+
 pub(crate) enum Storage {
     Memory { tokens: Mutex<JSONTokens> },
     Disk(DiskStorage),
+    Custom(Box<dyn TokenStorage>),
 }
 
 impl Storage {
     pub(crate) async fn set<T>(
         &self,
         scopes: ScopeSet<'_, T>,
         token: TokenInfo,
-    ) -> Result<(), io::Error>
+    ) -> anyhow::Result<()>
     where
         T: AsRef<str>,
     {
         match self {
-            Storage::Memory { tokens } => tokens.lock().await.set(scopes, token),
-            Storage::Disk(disk_storage) => disk_storage.set(scopes, token).await,
+            Storage::Memory { tokens } => Ok(tokens.lock().await.set(scopes, token)?),
+            Storage::Disk(disk_storage) => Ok(disk_storage.set(scopes, token).await?),
+            Storage::Custom(custom_storage) => {
+                let str_scopes: Vec<_> = scopes
+                    .scopes
+                    .iter()
+                    .map(|scope| scope.as_ref())
+                    .sorted()
+                    .unique()
+                    .collect();
```

Comment:
The allocation here is unfortunate. Comment for @dermesser: It may be worth considering changing the public API so that scope lists are passed as `&[&str]` rather than generic `&[T] where T: AsRef<str>`. I liked the latter when it was possible to do it without unnecessary overhead, but if it's causing the implementation to copy the scope list on every call to get a new token (as it now will when using a custom storage token) it's probably not worth the usability benefit.

Reply:
well, let's be realistic: We are dealing with -- sometimes intercontinental -- network calls here, over a HTTP2 connection with JSON. And then possibly seconds for an API response that is tens to hundreds of times bigger than the scope strings.

A couple allocations for short strings are acceptable here, IMHO, given the nice properties of `&[T] where T: AsRef<str>`.